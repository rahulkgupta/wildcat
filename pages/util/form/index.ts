// Form should be a class with a functionality like getField, etc.
// on change, react element should run field validation and field conditionals 
// form should keep all the state, so when a field is updated, the react element calls field.update. 



export class Field {
    id: string
    label: string
    error: string

    constructor(id: string, data: any) {
        this.id = id;
        this.label = data.label;
        this.error = data.error;
    }

    getLabel() {
        return this.label;
    }

    getError() {
        return this.error;
    }
}

export class TextField extends Field {

    constructor(id: string, data: any) {
        super(id, data);
    }

}

class FieldFactory {

    public createField(id: string, fieldData: any): Field {
        switch(fieldData.type) {
            case 'text': return new TextField(id, fieldData);
            default: throw Error('no field type found');
        }
    }
}

export default class Form {
    id: string
    factory: FieldFactory
    fields: Array<Field>

    constructor(data: any) {
        this.id = data.id;
        this.factory = new FieldFactory();
        this.fields = []
        for (const id in data.fields) {
            this.fields.push(this.factory.createField(id, data.fields[id]))
        }
    }

    getFields() {
        return this.fields;
    }
}