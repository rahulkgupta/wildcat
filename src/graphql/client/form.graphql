query getForm($id: uuid!) {
  forms(where: {id: {_eq: $id}}) {
    data
  }
}