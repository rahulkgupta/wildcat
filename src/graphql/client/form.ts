const getQuery = (id: string): string => {
  return `{
    forms_by_pk(id: "${id}") {
      data,
      id
    }
  }`;
};

export default getQuery;
