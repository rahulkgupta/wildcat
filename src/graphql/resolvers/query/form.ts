import { getFormByID } from '@src/db/datastore';
import { QueryResolvers } from '@src/graphql/sdk';

const Query: QueryResolvers = {
  form: (parent, args, context, info) => {
    return getFormByID(args.id);
  },
};

export default Query.form;
