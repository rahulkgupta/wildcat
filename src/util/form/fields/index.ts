

/**
 * Abstract Field class
 * Provides basic functionality for a Field
 * If you want to create a new type of Field, extend this class.
 */
export default abstract class Field {
    id: string // id of the field
    label: string // label of the field, such as "Enter your Name"
    error: string // error, used for when someone inputs something incorrectly
    onUpdate: Function
    
    /**
     * 
     * @param id id of the field
     * @param data data to use for the field. primarly `label` and `error`. Probably needs to be cleaned up.
     * @param onUpdate function used to call when a Field is updated as in {@link TextField.setValue}
     */
    constructor(id: string, data: any, onUpdate: Function) {
        this.id = id;
        this.label = data.label;
        this.error = data.error;
        this.onUpdate = onUpdate;
    }

    /**
     * @returns Field id.
     */
    getId() {
        return this.id;
    }
    
    /**
     * @returns Field label
     */
    getLabel() {
        return this.label;
    }

    /**
     * @returns Field error
     */
    getError() {
        return this.error;
    }

    /**
     * null placeholder for update function. Called in {@link Form.update}
     */
    update() {
        return null;
    }
}

