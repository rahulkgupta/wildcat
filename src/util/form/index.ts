import Field from "./fields";
import TextField from "./fields/text";

/**
 * Factory class that creates fields based on their type.
 */
class FieldFactory {
  /**
   *
   * @param id The id of the field
   * @param fieldData The data for the field. We use fieldData.type` to determine what field to instantiate.
   * @param onUpdate the function to call after a field updates.
   */
  public createField(id: string, fieldData: any, onUpdate: Function): Field {
    switch (fieldData.type) {
      case "text":
        return new TextField(id, fieldData, onUpdate);
      default:
        throw Error("no field type found");
    }
  }
}

interface Hooks {
  /**
   * This is called after the form updates. See {@link Form.update}
   */
  onUpdate: Function;
}

/**
 * This class is the main interface with React at the moment and contains the business logic.
 */
export default class Form {
  id: string; // id of the form
  factory: FieldFactory; // generates the Fields
  fields: Array<Field>; // array of fields
  onUpdate: Function; // passed in by the hooks argument

  /**
   * Instantiating a Form
   * @param data JSON data from the API (pages/api/forms/[fid].ts)
   * @param hooks This allows consumers of the Form to hook into the lifecycle of the form. See {@link Hooks} for more detail.
   */
  constructor(data: any, hooks: Hooks) {
    this.id = data.id;
    this.factory = new FieldFactory();
    this.fields = [];
    this.onUpdate = hooks.onUpdate;
    for (const id in data.fields) {
      this.fields.push(
        this.factory.createField(id, data.fields[id], this.update.bind(this))
      );
    }
  }

  /**
   * Updates the state of the form.
   * This function is passed into {@link FieldFactory.createField}
   * which is then passed to the Field constructor.
   */
  update() {
    this.fields.map((field) => field.update());
    this.onUpdate();
  }
  getFields() {
    return this.fields;
  }
}
