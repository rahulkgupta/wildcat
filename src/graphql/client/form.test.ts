import form from './form';

describe('form', () => {
  it('works', () => {
    expect(form('test')).toMatchSnapshot();
  });
});
