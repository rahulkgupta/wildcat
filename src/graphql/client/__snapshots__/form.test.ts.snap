// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form works 1`] = `
"{
    forms_by_pk(id: \\"test\\") {
      data,
      id
    }
  }"
`;
